src/main/java/org/opentripplanner/updater/trip/TimetableSnapshotManager.java

```diff
@@ -30,36 +29,18 @@ public final class TimetableSnapshotManager {
 
   private static final Logger LOG = LoggerFactory.getLogger(TimetableSnapshotManager.class);
   private final TransitLayerUpdater transitLayerUpdater;
-  /**
-   * Lock to indicate that buffer is in use
-   */
-  private final ReentrantLock bufferLock = new ReentrantLock(true);
 
   /**
-   * The working copy of the timetable snapshot. Should not be visible to routing threads. Should
-   * only be modified by a thread that holds a lock on {@link #bufferLock}. All public methods that
-   * might modify this buffer will correctly acquire the lock. By design, only one thread should
-   * ever be writing to this buffer.
-   * TODO RT_AB: research and document why this lock is needed since only one thread should ever be
-   *   writing to this buffer. One possible reason may be a need to suspend writes while indexing
-   *   and swapping out the buffer. But the original idea was to make a new copy of the buffer
-   *   before re-indexing it. While refactoring or rewriting parts of this system, we could throw
-   *   an exception if a writing section is entered by more than one thread.
+   * The working copy of the timetable snapshot. Should not be visible to routing threads.
+   * By design, only one thread should ever be writing to this buffer.
    */
   private final TimetableSnapshot buffer = new TimetableSnapshot();
 
   /**
    * The last committed snapshot that was handed off to a routing thread. This snapshot may be given
-   * to more than one routing thread if the maximum snapshot frequency is exceeded.
+   * to more than one routing thread.
    */
-  private volatile TimetableSnapshot snapshot = null;
-
-  /**
-   * If a timetable snapshot is requested less than this number of milliseconds after the previous
-   * snapshot, just return the same one. Throttles the potentially resource-consuming task of
-   * duplicating a TripPattern -> Timetable map and indexing the new Timetables.
-   */
-  private final CountdownTimer snapshotFrequencyThrottle;
+  private final ConcurrentPublished<TimetableSnapshot> snapshot = new ConcurrentPublished<>();
```

Comment:
I think I already know the answer but want to ask anyway: the use of `ConcurrentPublished` adds another layer of safety even though it isn't strictly necessary as there are never two writer threads active at the same time. Correct?

Or is there a case where its synchronization features are useful?

Reply:
Synchronization has 2 properties: _mutual exclusion_ and _safe publication_.
Mutual exclusion is not needed here since, like you said, there is only one writer thread and publishing the snapshot is an atomic operation (updating a reference).
What we need here is safe publication, that is: making sure that the reader thread sees all the modifications done by the writer thread. In practice it means purging the CPU caches and registers so that the reader thread does not use stale data.
Using a volatile field, as done in the current implementation, provides also a guarantee of safe publication, but it is a bit more obscure than using the `ConcurrentPublished` mechanism.

More details there: https://shipilev.net/blog/2014/safe-public-construction/

Reply:
The `ConcurrentPublished` is simple enough, but would the Java AtomicReference be a better alternative. There are pros/cons here, and in the end I think we will revisit this, so maybe leave it for now.

Reply:
My understanding is that AtomicReference is used when you need to do a compare-and-swap (check that the reference is the one you think it is before replacing it). Isn't the ConcurrentPublished synchronization doing something deeper here by imposing happens-before, and blocking reordering-based optimizations? 

An example of the problem I think it's avoiding: 
Code running on the single graph-writer thread changes some sub-object referenced by the snapshot, such as a TripTimes. It then updates the shared reference to the snapshot containing those TripTimes, but due to some reordering optimization (which would be harmless in a single threaded environment) the reference is published before the contents of the TripTimes sub-object is set. Even though setting a reference to the snapshot would always be atomic (no word-tearing when seen from other threads), there's no guarantee that the reading thread sees the new reference _after_ it sees changes to the TripTimes. Synchronization should establish a kind of reordering "fence" where everything appearing higher up in the source code is definitely visible to the reader.

We're talking about a place where there's likely to be read/write contention only once every 1-5 seconds so extremely rarely at the CPU-timescale. It's probably harmless to synchronize as ConcurrentPublished does, to get the benefit of strong ordering effects.

Reply:
@vpaturet checked the Javadoc in detail. For these purposes, volatile, AtomicReference (Javadoc on the `atomic` package rather than one class), and synchronization all work because they all establish happens-before relationships, and we don't have to deal with read/write contention here.